File: Java/22.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        backTrack(sb, 0, 0, n, result);
        return result;
    }

    private void backTrack(StringBuilder current, int open, int close, int n, List<String> result) {
        if (open == n && close == n) {
            result.add(current.toString());
            return;
        }

        if (open < n) {
            current.append('(');
            backTrack(current, open + 1, close, n, result);
            current.deleteCharAt(current.length() - 1);
        }

        if (close < open) {
            current.append(')');
            backTrack(current, open, close + 1, n, result);
            current.deleteCharAt(current.length() - 1);
        }
    }
}

File: Python/22.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        current = []

        def backtrack(open, close):
            if open == close == n:
                result.append("".join(current))
                return
            if open < n or open == close:
                current.append("(")
                backtrack(open + 1, close)
                current.pop()
            if open > close:
                current.append(")")
                backtrack(open, close + 1)
                current.pop()

        backtrack(0, 0)
        return result
